src/lib.rs

```diff
@@ -266,18 +266,31 @@ pub mod util {
     }
 
     pub fn load_private_key(filename: &str) -> Result<rustls::PrivateKey> {
-        let keyfile = fs::File::open(filename).map_err(|e| Error::Io(e))?;
-        let mut reader = BufReader::new(keyfile);
-        let mut keys = pemfile::rsa_private_keys(&mut reader)
-            .map_err(|_| Error::BadKey)?;
-
-        // Ensure there's only one key in the file.
-        if keys.len() != 1 {
-            return Err(Error::BadKeyCount);
+        use std::io::Seek;
+        let start = io::SeekFrom::Start(0);
+
+        fn expect_one(mut keys: Vec<rustls::PrivateKey>) -> Result<rustls::PrivateKey> {
+            if keys.len() != 1 {
+                Err(Error::BadKeyCount)
+            } else {
+                Ok(keys.remove(0))
+            }
         }
 
+        let keyfile = fs::File::open(filename).map_err(Error::Io)?;
+        let mut reader = BufReader::new(keyfile);
+        // Try reading one pkcs8 private key, or else one rsa private key
+        let key = pemfile::pkcs8_private_keys(&mut reader)
+            .map_err(|_| Error::BadKey)
+            .and_then(expect_one)
+            .or_else(|_| {
```

Comment:
This `or_else` will called if reading/parsing a PKCS#8 file succeeds but `expect_one` fails. We probably don't want to try to read the file again in this case.